mesonbuild/_pathlib.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import typing as T
+
+# Python 3.5 does not have the strict kwarg for resolve and always
+# behaves like calling resolve with strict=True in Python 3.6+
+#
+# This module emulates the behavior of Python 3.6+ by in Python 3.5 by
+# overriding the resolve method with a bit of custom logic
+#
+# TODO: Drop this module as soon as Python 3.5 support is dropped
+
+if T.TYPE_CHECKING:
+    from pathlib import Path
+else:
+    if sys.version_info.major <= 3 and sys.version_info.minor <= 5:
+
+        # Inspired by https://codereview.stackexchange.com/questions/162426/subclassing-pathlib-path
+        # from pathlib import Path as _Path_, _windows_flavour, _posix_flavour
+        import pathlib
+        import os
+
+        class Path(type(pathlib.Path())):
```

Comment:
Why this this not just `Path(pathlib.Path)`

Reply:
Because `pathlib.Path` is special... The `__new__` operator of `pathlib.Path` creates either a `PosixPath` or a `WindowsPath` object. Inheriting from `pathlib.Path` directly can work if you replicate the `__new__` operator which seems way messier to me.

Reply:
Ahhh, okay. I hadn't looked at the implementation of path lib, I assumed it was like os.path where Path was implemented in private modules and selectively imported. Can you add a comment here? With that I think this is good to go

Reply:
Done, thanks for the review.